src/onnx/parse_if.cpp

```diff
@@ -65,13 +67,95 @@ struct parse_if : op_parser<parse_if>
         auto else_out_shapes = else_mdl->get_output_shapes();
-        if(not std::equal(then_out_shapes.begin(),
-                          then_out_shapes.end(),
-                          else_out_shapes.begin(),
-                          else_out_shapes.end()))
+
+        auto throw_shapes = [&]() {
+            MIGRAPHX_THROW("PARSE_IF: " + info.name +
+                            " then and else sub_graphs must compatible shapes ");
+        };
+
+        if(then_out_shapes.size() != else_out_shapes.size())
+        {
+            throw_shapes();
+        }
+
+        // Must have the same type for both if/else blocks by onnx spec
+        // Add exception for empty constant scalars
+        if(then_out_shapes.at(0).type() != else_out_shapes.at(0).type())
         {
             MIGRAPHX_THROW("PARSE_IF: " + info.name +
-                           " then and else sub_grahps must have same output shapes!");
+                           " then and else sub_grahps must have same output type! " +
+                           then_out_shapes.at(0).type_string() + " vs " +
+                           else_out_shapes.at(0).type_string());
+        }
+
+        if(not then_out_shapes.at(0).dynamic() && not else_out_shapes.at(0).dynamic())
+        {
+            auto then_lens   = then_out_shapes.at(0).lens();
+            auto else_lens   = else_out_shapes.at(0).lens();
+
+            // Throw error if both branches have zero output shapes. Not possible for static inputs
+            if(then_lens.empty() && else_lens.empty())
+            {
+                throw_shapes();
+            }
+
+            auto handle_empty_branch = [](module_ref& mdl, const shape& out_shape) {
+                auto outline_ins = mdl->add_outline(out_shape);
+                mdl->replace_return({outline_ins});
+                return out_shape.lens();
+            };
+
+            // Handle one empty branch by setting output identical to the other
+            // need to update the then_shape before we do further checks
+            if(then_lens.empty())
+            {
+                then_lens = handle_empty_branch(then_mdl, else_out_shapes.at(0));
+            }
+            else if(else_lens.empty())
+            {
+                else_lens = handle_empty_branch(else_mdl, then_out_shapes.at(0));
+            }
+            else
+            {
+                // check equivilant length dims, and (x1,x2,.., xn, 1) == (x1,x2,..,xn)
+                int dim_delta = abs((static_cast<int>(then_lens.size() - else_lens.size())));
+
+                if(dim_delta <= 1)
+                {
+                    // make sure dims are equivalent in static shapes
+                    if(not equal(then_lens.begin(), then_lens.end(), else_lens.begin()) &&
+                       not equal(else_lens.begin(), else_lens.end(), then_lens.begin()))
+                    {
+                        throw_shapes();
+                    }
+
```

Comment:
I think this can seg fault:
ex: if then_lens.size() = 4 and else_lens.size() = 3 `equal(then_lens.begin(), then_lens.end(), else_lens.begin())` will check past the last value.

Reply:
Yeah just updated this. Was trying to be smart here and assumed equal played nice with ranges. Added a custom lambda to handle this nicely.